### PR TITLE
Added Python3 compatibility

### PR DESCRIPTION
**Pull request by lsqshr**
Added a few tweaks to make NeuroM compatible with both python2 and python3.

I tested this branch with both python2 and python3 `python setup.py test` & all test passed.

Pls consider a merge.

### CODE REVIEW

**Comment by mgeplf** [comment]
Hello,

Thanks for all the changes!

After a couple changes, I think we should be able to merge this.  Until we get continuous integration running w/ python3, I'm not sure that we'll be able to guarantee compatibility, but now that the bulk of the work is done, I will try and get py3 CI going.

Please bear with me, as I'm quite busy at the moment, so I'm not sure how fast I'll get to this.

**Comment by lsqshr** [comment]
That's great! Pls take your time. In the meanwhile we have this branch to be used with our python3 code & will contribute when we find anything.

File: examples/end_to_end_distance.py [hunk 1]
```diff
@@ -77,18 +77,18 @@ def _dist(seg):
     nrn = nm.load_neuron(filename)
 
     # print mean end-to-end distance per neurite type
-    print('Mean end-to-end distance for axons: ',
-          mean_end_to_end_dist(n for n in nrn.neurites if n.type == nm.AXON))
-    print('Mean end-to-end distance for basal dendrites: ',
-          mean_end_to_end_dist(n for n in nrn.neurites if n.type == nm.BASAL_DENDRITE))
-    print('Mean end-to-end distance for apical dendrites: ',
+    print(('Mean end-to-end distance for axons: ',
```
Comment (juanchopanza) [inline examples/end_to_end_distance.py]: Too many parentheses here.

File: examples/radius_of_gyration.py [hunk 2]
```diff
@@ -97,10 +97,10 @@ def mean_rad_of_gyration(neurites):
             radius_of_gyration(nrte), nrte.type) for nrte in nrn.neurites])
 
     # print mean radius of gyration per neurite type
-    print('Mean radius of gyration for axons: ',
-          mean_rad_of_gyration(n for n in nrn.neurites if n.type == nm.AXON))
-    print('Mean radius of gyration for basal dendrites: ',
-          mean_rad_of_gyration(n for n in nrn.neurites if n.type == nm.BASAL_DENDRITE))
-    print('Mean radius of gyration for apical dendrites: ',
+    print(('Mean radius of gyration for axons: ',
```
Comment (juanchopanza) [inline examples/radius_of_gyration.py]: Too many parens.

File: neurom/check/morphtree.py [hunk 3]
```diff
@@ -49,7 +52,7 @@ def is_monotonic(neurite, tol):
     for node in neurite.iter_sections():
         # check that points in section satisfy monotonicity
         sec = node.points
-        for point_id in xrange(len(sec) - 1):
+        for point_id in range(len(sec) - 1):
```
Comment (juanchopanza) [inline neurom/check/morphtree.py]: Does this give us `xrange` in python2?

File: neurom/check/morphtree.py [hunk 4]
```diff
@@ -173,7 +176,7 @@ def is_inside_cylinder(seg1, seg2):
         return not is_in_the_same_verse(seg1, seg2) and is_seg1_overlapping_with_seg2(seg1, seg2)
 
     # filter out single segment sections
-    for snode in ifilter(lambda snode: snode.points.shape[0] > 2, neurite.iter_sections()):
+    for snode in filter(lambda snode: snode.points.shape[0] > 2, neurite.iter_sections()):
```
Comment (juanchopanza) [inline neurom/check/morphtree.py]: Does this give us `ifilter` in python2?

File: neurom/check/neuron_checks.py [hunk 5]
```diff
@@ -33,15 +33,16 @@
...
-from itertools import izip
+import sys
 from neurom.core import Tree
 from neurom.core.types import NeuriteType
 from neurom.core.dataformat import COLS
 from neurom.morphmath import section_length, segment_length
 from neurom.check.morphtree import get_flat_neurites, get_nonmonotonic_neurites
 from neurom.fst import _neuritefunc as _nf
 from neurom.check import CheckResult
+if sys.version_info < (3, 0):
```
Comment (juanchopanza) [inline neurom/check/neuron_checks.py]: This kind of thing should be isolated in one place.

File: neurom/check/neuron_checks.py [hunk 6]
```diff
@@ -120,7 +121,7 @@ def has_all_nonzero_segment_lengths(neuron, threshold=0.0):
     bad_ids = []
     for sec in _nf.iter_sections(neuron):
         p = sec.points
-        for i, s in enumerate(izip(p[:-1], p[1:])):
+        for i, s in enumerate(zip(p[:-1], p[1:])):
```
Comment (juanchopanza) [inline neurom/check/neuron_checks.py]: Does this give us `izip` in python2?

File: neurom/check/neuron_checks.py [hunk 7]
```diff
@@ -190,7 +191,7 @@ def has_no_jumps(neuron, max_distance=30.0, axis='z'):
     bad_ids = []
     axis = {'x': COLS.X, 'y': COLS.Y, 'z': COLS.Z, }[axis.lower()]
     for s in _nf.iter_sections(neuron):
-        it = izip(s.points, s.points[1:])
+        it = zip(s.points, s.points[1:])
```
Comment (juanchopanza) [inline neurom/check/neuron_checks.py]: `izip` in python2?

File: neurom/check/runner.py [hunk 8]
```diff
@@ -34,12 +34,19 @@
...
 import logging
+from future.utils import iteritems
 from neurom.io.utils import get_morph_files
 from neurom.io import load_data
 from neurom.exceptions import ConfigError
 from neurom.fst import _core as fst_core
 from neurom.check import check_wrapper
 
+import sys
+if sys.version_info < (3, 0):
```
Comment (juanchopanza) [inline neurom/check/runner.py]: See previous comment. All these switches should happen in one place only.

File: neurom/check/tests/test_morphtree.py [hunk 9]
```diff
@@ -57,7 +58,7 @@ def _make_monotonic(neuron):
             sec = node.points
             if node.parent is not None:
                 sec[0][COLS.R] = node.parent.points[-1][COLS.R] / 2.
-            for point_id in xrange(len(sec) - 1):
+            for point_id in range(len(sec) - 1):
```
Comment (juanchopanza) [inline neurom/check/tests/test_morphtree.py]: is the python2 behaviour maintained?

File: neurom/check/tests/test_morphtree.py [hunk 10]
```diff
@@ -76,7 +77,7 @@ def fake_tree(init_radius, mode):
         radius = init_radius
 
         sec = []
-        for _ in xrange(5):
+        for _ in range(5):
```
Comment (juanchopanza) [inline neurom/check/tests/test_morphtree.py]: python2 behaviour?

File: neurom/core/__init__.py [hunk 11]
```diff
@@ -28,12 +28,15 @@
...
 
-from itertools import ifilter, imap, chain, izip
+import sys
+from itertools import chain
 from .tree import Tree
 from .types import NeuriteType
 from ._soma import Soma, make_soma, SomaError
 from ._neuron import Section, Neurite, Neuron
 from .population import Population
+if sys.version_info < (3, 0):
```
Comment (juanchopanza) [inline neurom/core/__init__.py]: We shouldn't do this all over the code.

File: neurom/core/__init__.py [hunk 12]
```diff
@@ -65,8 +68,8 @@ def iter_neurites(obj, mapfun=None, filt=None):
     neurites = ((obj,) if isinstance(obj, Neurite)
                 else (obj.neurites if hasattr(obj, 'neurites') else obj))
 
-    neurite_iter = iter(neurites) if filt is None else ifilter(filt, neurites)
-    return neurite_iter if mapfun is None else imap(mapfun, neurite_iter)
+    neurite_iter = iter(neurites) if filt is None else filter(filt, neurites)
```
Comment (juanchopanza) [inline neurom/core/__init__.py]: python2 behaviour?

File: neurom/check/morphtree.py (hunk 4 above)
Comment (mgeplf) [inline neurom/check/morphtree.py]: It does from the:
```
if sys.version_info <= (3, 0):
    from itertools import ifilter as filter
```
at the top.

I think, however, we should consider moving all those checks into a compat.py, and have that use six.py.

File: neurom/check/morphtree.py (hunk 3 above)
Comment (mgeplf) [inline neurom/check/morphtree.py]: no, so this is possibly expensive, depending on len(sec)

Comment (mgeplf) [inline neurom/check/morphtree.py]: there's ```from builtins import range```
I do need to make sure that handles python2.6, though

**Comment by mgeplf** [comment]
Ok, I'm going through this now.  I'm following my plan of moving all necessary functionality to _compat.py, and going from there.

**Comment by mgeplf** [comment]
I couldn't push to your branch, so I opened another pull request here: #550.

I also am trying to get CI for python3, and python2.6 at the same time.  I will probably have to break these commits up for them to make sense, but I don't have the time now.  In addition, I am booked until next Friday with other stuff, so I won't be able to look further into this until the 25th.

Thanks

**Comment by lsqshr** [comment]
Good idea to have a `_compat.py` to do all the version checking in one place. Let's follow this up in #550

**Comment by mgeplf** [comment]
I'm back.

I'm going to close this up, and finish it in #550.

**Pull request closed without merging**